internal/daemon/controller/handlers/targets/tcp/grants_test.go

```diff
@@ -103,12 +178,407 @@ func TestGrants_ReadActions(t *testing.T) {
+		wantOutputFields []string
+	}
+	t.Run("Read", func(t *testing.T) {
+		testcases := []struct {
+			name               string
+			userFunc           func() (*iam.User, auth.Account)
+			wantIdOutputFields map[string]readTestResult
+		}{
+			{
+				name: "global role grant descendants returns all created targets",
+				userFunc: iam.TestUserDirectGrantsFunc(t, conn, kmsCache, globals.GlobalPrefix, password.TestAuthMethodWithAccount, []iam.TestRoleGrantsRequest{
+					{
+						RoleScopeId: globals.GlobalPrefix,
+						Grants:      []string{"ids=*;type=target;actions=list,read;output_fields=id,authorized_actions,scope_id,address"},
+						GrantScopes: []string{globals.GrantScopeDescendants},
+					},
+				}),
+				wantIdOutputFields: map[string]readTestResult{
+					target1.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.AuthorizedActionsField, globals.ScopeIdField, globals.AddressField}},
+					target2.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.AuthorizedActionsField, globals.ScopeIdField, globals.AddressField}},
+					target3.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.AuthorizedActionsField, globals.ScopeIdField, globals.AddressField}},
+				},
+			},
+			{
+				name: "org role grant this returns all created targets",
+				userFunc: iam.TestUserDirectGrantsFunc(t, conn, kmsCache, globals.GlobalPrefix, password.TestAuthMethodWithAccount, []iam.TestRoleGrantsRequest{
+					{
+						RoleScopeId: org1.GetPublicId(),
+						Grants:      []string{"ids=*;type=target;actions=list,read;output_fields=id,authorized_actions,scope_id,address"},
+						GrantScopes: []string{globals.GrantScopeChildren},
+					},
+				}),
+				wantIdOutputFields: map[string]readTestResult{
+					target1.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.AuthorizedActionsField, globals.ScopeIdField, globals.AddressField}},
+					target2.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.AuthorizedActionsField, globals.ScopeIdField, globals.AddressField}},
+					target3.GetPublicId(): {wantErr: handlers.ForbiddenError()},
+				},
+			},
+			{
+				name: "project role grant this returns all created targets",
+				userFunc: iam.TestUserDirectGrantsFunc(t, conn, kmsCache, globals.GlobalPrefix, password.TestAuthMethodWithAccount, []iam.TestRoleGrantsRequest{
+					{
+						RoleScopeId: proj2.GetPublicId(),
+						Grants:      []string{"ids=*;type=target;actions=list,read;output_fields=id,name,created_time,updated_time,version"},
+						GrantScopes: []string{globals.GrantScopeThis},
+					},
+				}),
+				wantIdOutputFields: map[string]readTestResult{
+					target1.GetPublicId(): {wantErr: handlers.ForbiddenError()},
+					target2.GetPublicId(): {wantErr: handlers.ForbiddenError()},
+					target3.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.NameField, globals.CreatedTimeField, globals.UpdatedTimeField, globals.VersionField}},
+				},
+			},
```

Comment:
Can we add a case for an individual GrantScope?
```suggestion
			},
			{
				name: "global role with individual project grant returns all created targets on the project",
				userFunc: iam.TestUserDirectGrantsFunc(t, conn, kmsCache, globals.GlobalPrefix, password.TestAuthMethodWithAccount, []iam.TestRoleGrantsRequest{
					{
						RoleScopeId: globals.GlobalPrefix,
						Grants:      []string{"ids=*;type=target;actions=list,read;output_fields=id,name,created_time,updated_time,version"},
						GrantScopes: []string{proj2.PublicId},
					},
				}),
				wantIdOutputFields: map[string]readTestResult{
					target1.GetPublicId(): {wantErr: handlers.ForbiddenError()},
					target2.GetPublicId(): {wantErr: handlers.ForbiddenError()},
					target3.GetPublicId(): {wantOutputFields: []string{globals.IdField, globals.NameField, globals.CreatedTimeField, globals.UpdatedTimeField, globals.VersionField}},
				},
			},
```